File: Linked List/L8_ciecular_linklist_all_operations.cpp
#include<iostream>
using namespace std;
// structure of circular linklist
struct Node{
    int data;
    Node *next;
    Node(int d)
    {
        data=d;
        next=NULL;
    }
};
// Circulat linklist traversal
// Method 1
void printList(Node *head)
{
    if(head==NULL) return;
    Node *curr=head;
    do
    {
        cout<<(curr->data)<<" ";
        curr=curr->next;
    } while (curr!=head);
    cout<<endl;
}
// Insert at begining of circular linklist: Method 1
// Naive method O(n)
Node *insertBegin(Node *head,int x)
{
    // create a node value x
    Node *temp=new Node(x);
    // If no element is present in circular linklist
    if(head==NULL)
        temp->next=temp;
    else
    {
        // if at least one element present
        Node *curr=head;
        while(curr->next!=head)
        {
            curr=curr->next;
        }
        curr->next=temp;
        temp->next=head;
        
    }
    return temp;
}
// Effective method 
// Insert at begining of circular linklist: Method 2
// insert new element at 2nd position 
// then swap 1st and 2nd element
// time =O(1)
Node *insertBegin2(Node *head, int x)
{
    Node *temp=new Node(x);
    if(head==NULL)
    {
        temp->next=temp;
        return temp;
    }
    else
    {
        // insert item at 2nd poistion(below 2 linke code)
        temp->next=head->next;
        head->next=temp;
        // swap the item
        int t=head->data;
        head->data=temp->data;
        temp->data=t;

        return head;    // returning old head(no change)
    }
}
// Insert at the end of Circular Linked List
// time 𝛳(n)
Node *insertEnd(Node *head,int x)
{
    Node *temp=new Node(x);
    if(head==NULL)
    {
        temp->next=temp;
        return temp;
    }
    else
    {
        Node *curr=head;
        while(curr->next!=head)
        {
            curr=curr->next;
        }
        curr->next=temp;
        temp->next=head;
        return head;
    }
}
// Effective method
// insert at end
// insert new element at 2nd position 
// then swap 1st and 2nd element
// take head pointer at 2nd position
// time =O(1)
Node *insertEnd2(Node *head, int x)
{
    Node *temp=new Node(x);
    if(head==NULL)
    {
        temp->next=temp;
        return temp;
    }
    else
    {
        // insert item at 2nd poistion(below 2 line code)
        temp->next=head->next;
        head->next=temp;
        // swap the item
        int t=head->data;
        head->data=temp->data;
        temp->data=t;
        head=head->next;    // keep head pointer at 2nd position
        return head;    // returning old head(no change)
    }
}
// Delete Head of Circular Linked List
// Naive mathod
// time = O(n)
Node *deleteHead(Node *head)
{
    if(head==NULL) return NULL;
    if(head->next==NULL)
    {
        delete head;
        return NULL;
    }
    else
    {
        Node *curr=head;    // curr pointing head node
        Node *tail=head;
        while(tail->next!=head)
        {
            tail=tail->next;    
        }
        // Now tail pointer reach at last node
        head=head->next;    // keep head at 2nd position
        tail->next=head;    // connecting tail to head
        curr->next=NULL;    // disconnecting curr node(previously head)
        delete curr;
        return head;


        // other method
        // we can solve only using one extra pointer
        /*
        Node *tail=head;
        while(tail->next!=head)
        {
            tail=tail->next;
        }
        tail->next=head->next;
        delete head;
        return (tail->next); */
    }
}
// Delete Head of Circular Linked List
// Effective method
// time = O(1)
Node *deleteHead2(Node *head)
{
    if(head==NULL) return NULL;
    if(head->next==NULL)
    {
        delete head;
        return NULL;
    }
    else
    {
        Node *curr=head;
        curr=curr->next;
        //swap data
        int t=head->data;
        head->data=curr->data;
        curr->data=t;
        // delete curr node
        head->next=curr->next; // link head(new head) with next of curr node 
        curr->next=NULL;    // disconnect curr node
        delete curr;
        return head;
    }    
}
// Delete Kth of a Circular Linked List
Node *delKthElement(Node *head, int k)
{
    if(head==NULL) return NULL;
    if(head->next==NULL)
    {
        delete head;
        return NULL;
    }
    else
    {
        Node *curr=head;
        Node *target=head;
        int count=1;
        while(count<k-1)
        {
            curr=curr->next;
            target=target->next;
            count++;
        }
        target=target->next;

        curr->next=target->next;
        target->next=NULL;
        delete target;
        return head;
    }
}
// main function
int main(){
    Node *head=NULL;
    head=insertBegin(head,15);
    head=insertBegin(head,5);
    head=insertBegin(head,20);
    head=insertBegin(head,18);
    cout<<"insert at begining(Naive method):"<<endl;
    printList(head);

    head=insertBegin2(head,11);
    cout<<"insert at begining(Effective method):"<<endl;
    printList(head);

    cout<<"insert 100 at End(Naive method):"<<endl;
    head=insertEnd(head,100);
    printList(head);

    cout<<"insert 90 at End(Effective method):"<<endl;
    head=insertEnd2(head,90);
    printList(head);

    cout<<"After deleting head(Naive method):"<<endl;
    head=deleteHead(head);
    printList(head);

    cout<<"After deleting head(Effective method):"<<endl;
    head=deleteHead2(head);
    printList(head);

    cout<<"After deleting 3rd element(15):"<<endl;
    head=delKthElement(head,3);
    printList(head);
    return 0;
}

File: code_java/self pased DSA/Linked List/L8_ciecular_linklist_all_operations.java
/**
 * Problem Statement:
 * Implement a Circular Singly Linked List in Java supporting the following operations:
 * - Insert at the beginning (naive and efficient methods).
 * - Insert at the end (naive and efficient methods).
 * - Delete the head node (naive and efficient methods).
 * - Delete the Kth node.
 * - Print the list.
 *
 * Approach:
 * - Use a static nested Node class to represent each node.
 * - Maintain circular links while performing insertions and deletions.
 * - Use data swapping for efficient O(1) insertions/deletions.
 * - Use modular, static methods to perform operations.
 *
 * Time Complexity:
 * - Insert at Beginning (Naive): O(n)
 * - Insert at Beginning (Efficient): O(1)
 * - Insert at End (Naive): O(n)
 * - Insert at End (Efficient): O(1)
 * - Delete Head (Naive): O(n)
 * - Delete Head (Efficient): O(1)
 * - Delete Kth Node: O(k)
 * - Print List: O(n)
 *
 * Space Complexity:
 * - O(1) auxiliary space (no additional structures used).
 */
// Circular Linked List implementation in Java with all operations

class CircularLinkedList {
    static class Node {
        int data;
        Node next;

        Node(int d) {
            data = d;
            next = null;
        }
    }

    // Print circular linked list
    static void printList(Node head) {
        if (head == null) return;
        Node curr = head;
        do {
            System.out.print(curr.data + " ");
            curr = curr.next;
        } while (curr != head);
        System.out.println();
    }

    // Naive insert at beginning
    static Node insertBegin(Node head, int x) {
        Node temp = new Node(x);
        if (head == null) {
            temp.next = temp;
            return temp;
        }
        Node curr = head;
        while (curr.next != head) {
            curr = curr.next;
        }
        curr.next = temp;
        temp.next = head;
        return temp;
    }

    // Efficient insert at beginning
    static Node insertBegin2(Node head, int x) {
        Node temp = new Node(x);
        if (head == null) {
            temp.next = temp;
            return temp;
        }
        temp.next = head.next;
        head.next = temp;
        int t = head.data;
        head.data = temp.data;
        temp.data = t;
        return head;
    }

    // Naive insert at end
    static Node insertEnd(Node head, int x) {
        Node temp = new Node(x);
        if (head == null) {
            temp.next = temp;
            return temp;
        }
        Node curr = head;
        while (curr.next != head) {
            curr = curr.next;
        }
        curr.next = temp;
        temp.next = head;
        return head;
    }

    // Efficient insert at end
    static Node insertEnd2(Node head, int x) {
        Node temp = new Node(x);
        if (head == null) {
            temp.next = temp;
            return temp;
        }
        temp.next = head.next;
        head.next = temp;
        int t = head.data;
        head.data = temp.data;
        temp.data = t;
        return temp;
    }

    // Naive delete head
    static Node deleteHead(Node head) {
        if (head == null || head.next == head) return null;
        Node curr = head;
        while (curr.next != head) {
            curr = curr.next;
        }
        curr.next = head.next;
        head.next = null;
        return curr.next;
    }

    // Efficient delete head
    static Node deleteHead2(Node head) {
        if (head == null || head.next == head) return null;
        Node curr = head.next;
        int t = head.data;
        head.data = curr.data;
        curr.data = t;
        head.next = curr.next;
        return head;
    }

    // Delete Kth element
    static Node delKthElement(Node head, int k) {
        if (head == null) return null;
        if (k == 1) return deleteHead(head);

        Node curr = head;
        for (int i = 0; i < k - 2; i++) {
            curr = curr.next;
        }
        Node temp = curr.next;
        curr.next = temp.next;
        return head;
    }

    public static void main(String[] args) {
        Node head = null;

        head = insertBegin(head, 15);
        head = insertBegin(head, 5);
        head = insertBegin(head, 20);
        head = insertBegin(head, 18);
        System.out.println("insert at beginning (Naive):");
        printList(head);

        head = insertBegin2(head, 11);
        System.out.println("insert at beginning (Efficient):");
        printList(head);

        head = insertEnd(head, 100);
        System.out.println("insert 100 at end (Naive):");
        printList(head);

        head = insertEnd2(head, 90);
        System.out.println("insert 90 at end (Efficient):");
        printList(head);

        head = deleteHead(head);
        System.out.println("after deleting head (Naive):");
        printList(head);

        head = deleteHead2(head);
        System.out.println("after deleting head (Efficient):");
        printList(head);

        head = delKthElement(head, 3);
        System.out.println("after deleting 3rd element:");
        printList(head);
    }
}